### PR TITLE
Add initial Treadmill Hardware CI workflow scaffold

### PR DESCRIPTION
**Pull request by lschuermann**
### Pull Request Overview

This pull request adds a basic GitHub actions workflow that kicks off a job on the [Treadmill Hardware Testbed](https://book.treadmill.ci) for every PR, merge, and push to master. It consists of two stages:
- `test-prepare`:  runs on the hosted GitHub Actions infrastructure and uses the GitHub API and Treadmill CLI to create a new, ephemeral, just-in-time GitHub actions runner that is then started on a Treadmill host, and
- `test-execute`: runs on the Treadmill infrastructure, with full access to the hardware boards connected to the platform.

Notably, the `test-prepare` step can, in the future, analyze the changeset of the PR in question to identify a set of subsystems and create a set of Treadmill jobs on boards that feature these subsystems. For this, the workflow file can run multiple `test-execute` stages in parallel.

Currently, the `test-execute` stage does not do anything useful -- it merely compiles the Tock kernel and lists available USB devices. A follow-up PR by @charles37 will introduce a Python script that uses this infrastructure to perform a set of automated integration tests.

I documented the entire setup process to integrate the Treadmill testbed with the Tock repository here; these steps have already been completed: https://book.treadmill.ci/user-guide/github-actions-integration.html

### Testing Strategy

This pull request tests itself. It might need to be merged to be picked up by GitHub.

### TODO or Help Wanted

We fully expect these things to break. This is the first "production" test of the Treadmill software, and there are still bugs. Thus, I advocate to leave this as an optional check for the near future. Having this PR integrated into the Tock repository and run checks on each PR, merge, and push to master will allow us to fix the remaining bugs and provides valuable experience in running this system.

~This PR is not quite ready yet -- we're still working on marking jobs as terminated once the hypervisor exists or the netbooted targets request a shutdown. The current, third `test-terminate` job is brittle and does not run in case of failures. We expect to have this done in the next few hours / days.~

The Treadmill job now automatically terminates once the GitHub Actions runner has processed the Actions job.

### Documentation Updated

We're writing documentation of the Treadmill infrastructure and testbed itself in a dedicated mdBook: https://book.treadmill.ci/

Once this infrastructure runs actual test workloads, we should add Tock-specific documentation to its mdBook as well.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

**Comment by lschuermann** [comment]
It looks like the workflow runs through successfully: https://github.com/tock/tock/actions/runs/10995587055/job/30526837543

One thing that's a little annoying with this is that we currently need to approve every run of this workflow. This is because, to create a new "just-in-time" GitHub actions runner, we have to use GitHub's API with elevated privileges. Specifically, as part of the `test-prepare` workflow, we have to use the "Repository administration" permission to use this endpoint, and as the name implies, this permission can do a lot of damage.

I have created a GitHub "deployment environment" to store the secret key for the GitHub App that has this permission, and for any workflow to have access to these permissions, it now requires explicit approval from the Tock core WG. We just need to make sure that PRs that have this workflow run on them don't maliciously modify the `test-prepare` stage of the `treadmill-ci` workflow file, similar to how we already have to make sure that PRs don't try to sneak in new permissions into our existing workflow files.

As a consequence, we may need to switch this workflow to only run on merges, though. My hopes would be that GitHub can collect these approvals upon clicking "Merge with ready", so we won't be bothered with it after that.

File: .github/workflows/treadmill-ci.yml
Comment (bradjc) [inline .github/workflows/treadmill-ci.yml:3]:
```suggestion
# Copyright Tock Contributors 2024.
```

**Review by bradjc — COMMENTED** [review]
Cool!

I think we do need to either write or link to documentation so that anyone who sees this file in our repo or related actions on a PR knows what treadmill is. Since treadmill isn't going to be google-able at the moment I think we need to be proactive in pointing people in the right direction.

**Comment by lschuermann** [comment]
I changed the workflow to use a different deploy environment for pushes to the master branch and GitHub merge queue refs -- whether the latter works is to be seen. In theory this should mean that we can run hardware tests on every merge and every push to master without approval.

By default, for pushes to a PR, GitHub will ask for an approval for _every_ push. This is deliberate, as the workflow file could have been changed between pushes, and looks like this:
image
We can approve this message (up to 30 days after the push) to launch the deployment (by clicking on the link under "requested a deployment".

**Comment by lschuermann** [comment]
This should be ready now!

**Review by bradjc — COMMENTED** [review]
Looks good for now!

>Currently, the test-execute stage does not do anything useful -- it merely compiles the Tock kernel and lists available USB devices. A follow-up PR by @charles37 will introduce a Python script that uses this infrastructure to perform a set of automated integration tests.

I think it makes sense to wait for that PR to review more thoroughly.

As a preview, I think we need to include pointers in the .yml file and a new doc in tock/docs that explains what is being run by treadmill and how to reproduce (in some form) the tests locally. Tracing through what is running on github actions can be tricky. But it makes sense to add those docs in the follow up PR.

**Comment by brghena** [comment]
@lschuermann I'm not really following what the status is on the Permissions issue here. Have you come up with some reasonable resolution? From skimming it over, I'm under the impression that a human has to click a button to run this every time, which doesn't feel reasonable (and is hopefully incorrect).

**Review by bradjc — APPROVED** [review]

**Review by ppannuto — APPROVED** [review]
> I'm advocating for this PR to be merged now, before the follow-up PR is created or merged, as it allows us to see whether this integration works at all, and whether we need to tweak it. The semantics as are exposed to the Tock repo will stay the same regardless of the follow up PR: this PR already adds the Actions workflow, approval, and status reporting, whereas the follow-up PR will simply change this workflow to run some actual tests.

Given the tight binding to GitHub infrastructure, I'm fine with merging rougher things here [especially as non-req'd checks].

That said, could parts of this (especially e.g. exploring ideas around the permissions stuff) be tested / iterated faster on a dedicated test repo you own that you hook treadmill to so you aren't blocked on Tock's review process for developing CI infra?

**Comment by alevy** [comment]
> could parts of this (especially e.g. exploring ideas around the permissions stuff) be tested / iterated faster on a dedicated test repo you own

I think that's actually happening, but some of the interactions are just very labor intensive to mimick, since they require things like other people opening PRs, users with different permissions, rebases and merges, etc.

**Comment by lschuermann** [comment]
> > could parts of this (especially e.g. exploring ideas around the permissions stuff) be tested / iterated faster on a dedicated test repo you own
> 
> I think that's actually happening, but some of the interactions are just very labor intensive to mimick, since they require things like other people opening PRs, users with different permissions, rebases and merges, etc.

Exactly. It's very hard to predict any issues we'll encounter based on our simulated interactions with the `treadmill-tb/tock` fork of the repository that we used to develop this.

> I'm not really following what the status is on the Permissions issue here. Have you come up with some reasonable resolution? From skimming it over, I'm under the impression that a human has to click a button to run this every time, which doesn't feel reasonable (and is hopefully incorrect).

That's a correct assessment, at least right now. We're working against somewhat fundamental issues with GitHub's API token permissions and the current architecture of these workflows now.

There's three options that we can mix and match:
- run HW CI on every push to an open PR: requires manual approval by the core WG & adds annoying "requests deployment" approval messages to the PR.
- run HW CI on PR being merged & master branch: requires no approval, as we can trust the workflow file.
- run HW CI on demand for any branch: requires manual action by people with the appropriate permissions & approval by Core WG.

In the short term, I'd suggest that we disable hardware CI checks for _every push to an arbitrary branch_ and _every PR_, which are the two problematic cases: because users control the GitHub actions workflow files, and those workflows have access to an API token with elevated privileges, they could be abused by attackers. This currently works around these issues by giving access to the secrets only after a manual approval by the Core WG.

If we want to test a PR on Treadmill before enqueuing it to merge, we can add the ability to launch & approve this workflow manually, on demand.

In the long term, we can work around these issues by hosting a central GitHub App that owns this token and performs the necessary API calls as a response to GitHub webhooks. That's yet another (small) piece of infrastructure we just haven't gotten around to building yet.

**Comment by jrvanwhy** [comment]
I've done a "perform privileged GitHub operations in response to incoming PRs without requiring approvals" thing before. I don't have the time needed to determine whether the same approach will work for Treadmill, so here it is, maybe this is helpful?

**Workflow A** is configured to run on every PR with no write permissions:
```yaml
name: A
on:
  pull_request:
permissions: {}
```

**Workflow B** is configured to run when workflow A completes, and runs in the default branch (`master`/`main`/etc):
```yaml
name: B
on:
  workflow_run:
    workflows: [A]
    types: [completed]
permissions:
  # Scary permissions here
```

When a PR is opened, workflow A runs, then workflow B runs. Workflow A can run untrustworthy code (e.g. checkout the PR's code and run the build system), while workflow B can do privileged operations. Workflow B is executed from the default branch, so an incoming PR cannot tamper with it. Workflow A can upload artifacts which workflow B downloads, allowing A to pass data to B.

The one thing that doesn't work that I would like to have working in workflow B does not show up in check list at the bottom of the PR -- to see its execution you have to navigate away from the PR to the actions tab (and if there are multiple PRs open at the same time, it's not clear which execution of workflow B corresponds to which PR).

If you think this may work for Treadmill, let me know, and I can send a link to the workflows with pointers for what to look at.

**Comment by lschuermann** [comment]
> The one thing that doesn't work that I would like to have working is workflow B does not show up in check list at the bottom of the PR -- to see its execution you have to navigate away from the PR to the actions tab (and if there are multiple PRs open at the same time, it's not clear which execution of workflow B corresponds to which PR).

Unfortunately I already explored this option and found this exact limitation. If we needed workflow B to have elevated privileges and perform the required and hopefully infallible interactions with GitHub's APIs, this would work. However, in our case, we have workflow A perform the privileged tasks, whereas workflow B contains the actual testing workload. Not having this second workflow related to the original trigger seems like a deal-breaker and makes depending on its as part of merge-queue checks etc. impossible.

We can use the `pull_request_target` trigger to have the workflow-file be taken from the main repository branch and also have access to secrets. However, that implies that all jobs in this workflow have elevated privileges as well (such as write access to the repository), and GitHub's guide cautions against running anything untrusted as part of these workflows. Thus this also doesn't seem like a {secure, satisfying} option unfortunately.

**Comment by bradjc** [comment]
I think we want this no matter what, and any manual issues with running a check are smaller than manually flashing a board.

Should we just merge this?

**Review by brghena — APPROVED** [review]

**Pull request merged**
